File: c++/22869.cpp
#include <iostream>

using namespace std;

int N, K;
int arr[5000];
bool dp[5000];

int main() {
  freopen("example.txt", "r", stdin);
  cin >> N >> K;

  for (int i = 0; i < N; i++) {
    cin >> arr[i];
  }

  dp[0] = true;

  for (int i = 0; i < N; i++) {
    for (int j = i + 1; j < N; j++) {
      if (dp[i] && ((j - i) * (1 + abs(arr[i] - arr[j]))) <= K) {
        dp[j] = true;
      };
    }
  }

  cout << (dp[N - 1] ? "YES" : "NO");

  return 0;
}

File: js/22869.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const [N, K] = inputs[0].split(' ').map(Number);
const arr = inputs[1].split(' ').map(Number);

const dp = Array(N).fill(false);
dp[0] = true;

for (let i = 0; i < N; i++) {
  for (let j = i + 1; j < N; j++) {
    if (dp[i] && calcPower(i, j) <= K) {
      dp[j] = true;
    }
  }
}

console.log(dp[N - 1] ? 'YES' : 'NO');

function calcPower(i, j) {
  return (j - i) * (1 + Math.abs(arr[i] - arr[j]));
}
